woocommerce/woocommerce
# Update restock logic during refunds closes #29502

### All Submissions:






### Changes proposed in this Pull Request:



This was worked on along with the help of @vedanshujain. We basically introduced a new meta to keep track of the count of stocks that are refunded and are opt in to restock. With this new count, we're able to get the difference in order to know how much stock to actually add or subtract from the line item whenever the "maybe_restock" method is called. We've decided to implement this fix for all new orders (WC 5.5+). Any orders before that will behave the same way as before.

Closes #29502 

### How to test the changes in this Pull Request:

PREREQ: run `composer dump-autoload`

1. Set https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-woocommerce.php#L27 to version 5.5
2. Create an order with an item that keeps stocks.
3. Go to the order and do a refund on the line item reducing the stock by 1. Ensure to uncheck the box that says `Restock refunded items`.  Click the `Refund manually` button.
4. Note stock count for the product stays the same and not restocked.
5. On the same order click the `Update` button. Ensure stock count remains the same for the product and not restocked.
6. Manually edit the quantity of the product line item. Reduce it by 1. Click on `Save` button on the bottom. Ensure stock is increased for the product.
7. Click on `Update` button. Ensure stock count remains the same for the product and not restocked.
8. Manually edit the quantity of the product line item again and this time increase it by 1. Click on `Save` button on the bottom. Ensure stock is decreased for the product.
9. Click on `Update` button. Ensure stock count remains the same for the product and not restocked.

### Other information:




### Changelog entry

> Fix - Introduce meta to track stocks that refunded and restocked to properly handle stock recalculation.
